### DISCUSSION TITLE
usb.util.find_descriptor return NoneType

### DISCUSSION THREAD

**Discussion by JianmingWANG1998** [pyusb/pyusb#428] [Q&A] [answered]
![image](https://user-images.githubusercontent.com/69441947/172564197-902d813e-c864-4bc4-9d71-5efa829a092c.png)
![image](https://user-images.githubusercontent.com/69441947/172564363-b4b30a5b-80d6-4e1f-8c0f-27036a065415.png)

**Comment by JianmingWANG1998** [comment]
Could you help me check why the function returns NoneType? I have followed the instruction on the github and fixed the errors mentioned in the FAQ. Also, I used the Zadig to install WinUSB. It also didn't work. 
![image](https://user-images.githubusercontent.com/69441947/172565132-61da3398-ee06-4e5a-9007-484ea4372b2c.png)
![image](https://user-images.githubusercontent.com/69441947/172565543-2e4d30db-c65d-4414-95ca-24627fb2963d.png)

**Comment by jonasmalacofilho** [comment]
As show in the output from `print(intf)`, that particular interface only has a single **IN** endpoint.

**Reply by JianmingWANG1998** [reply]
Thank you! You are right. The first interface only has a IN endpoint. After I choose the second interface, it works.

**Reply by JianmingWANG1998** [reply]
But I used the second interface to write "test" into the Bulk Out endpoint, it shows the NotImplementedError. Could you help me again? Really thank you for your help!
![image](https://user-images.githubusercontent.com/69441947/172748433-b51a5c22-4fb9-469a-b0e7-42bb8ac1eb8d.png)
![image](https://user-images.githubusercontent.com/69441947/172748664-e9a0d2e7-cb50-45be-a783-29c384e56f58.png)
![image](https://user-images.githubusercontent.com/69441947/172748697-fff810f4-4c6d-451a-8743-2ab2b20a8c11.png)
![image](https://user-images.githubusercontent.com/69441947/172748737-55b47811-dcd4-4b6b-8db9-e689580ef355.png)

**Reply by jonasmalacofilho** [reply, marked as answer]
It seems that with composite devices you may need to attach WinUSB (with Zadig) to the _parent_ device, and not only to a specific interface, like it seems you've done here (from your previous screenshot).[^1][^2]

P.S. In these situations you'll want to look into which specific call failed (`_LibUSB.claim_interface`, which is our wrapper for `libusb_claim_interface`),[^3] and what might cause LibUSB to return that error code (`LIBUSB_ERROR_NOT_SUPPORTED`).  Their documentation is the best place to start, but also take a look at their wiki and, failing that, their issue tracker and mailing list.

[^1]: https://github.com/pyusb/pyusb/issues/370#issuecomment-860195373
[^2]: https://github.com/libusb/libusb/issues/445#issuecomment-397302269
[^3]: https://libusb.sourceforge.io/api-1.0/group__libusb__dev.html#gaee5076addf5de77c7962138397fd5b1a

**Reply by JianmingWANG1998** [reply]
Thank you for your help! Really appreciate.

**Reply by JianmingWANG1998** [reply]
By following the above instruction, I have succeed to write "test1" into the device. However, when I try to read the information from the device, it shows USBTimeout Error. By using the nRF Connect android application to check whether the data is send to the device or not, the app shows that the data is successfully send to the device. So, I don't know why there is USBTimeout Error. I also try to make the timeout be larger like timeout = 20000. It still shows the error. Could you help me one more time? Really appreciate!
![image](https://user-images.githubusercontent.com/69441947/173287356-08e41c85-fe45-41fa-8841-1abfd1847864.png)
![image](https://user-images.githubusercontent.com/69441947/173287668-44ddc382-a0b2-464e-bd2c-fe027c01ff5f.png)
![image](https://user-images.githubusercontent.com/69441947/173287699-556e0b8e-72a1-406a-9f26-6132b01b5cfa.png)
![image](https://user-images.githubusercontent.com/69441947/173287739-d7b9264a-66a0-4c7c-948d-7d4f57b0b574.png)

**Reply by jonasmalacofilho** [reply]
I don't know how the CDC class protocol works, but timeouts during reads sometimes mean that the device is not supposed to reply anything at that time.  This is likely a (not-knowning-the-)protocol error, and you'll need to delve into the relevant specs and/or reverse engineer it.

There's (of course) also a small chance of a bug in PyUSB or LibUSB, especially since this is isn't (AFAIK) an interface class commonly used with these libraries.  But we have no chance of investigating that possibility before you're certain that you got the protocol right and, preferably, also have a comparable implementation using different libraries that works.
